v23.1/transaction-retry-error-reference.md

```diff
@@ -1,54 +1,96 @@
 
-- Why the error is happening.
-- What to do about it.
+## Overview
 
-This page is meant to provide information about specific transaction retry error codes to make troubleshooting easier. In most cases, the correct actions to take when these errors occur are:
+CockroachDB always attempts to find a [serializable ordering](demo-serializable.html) among all of the currently-executing transactions.
 
-1. Update your app to retry on serialization errors (where `SQLSTATE` is `40001`), as described in [client-side retry handling](transactions.html#client-side-intervention).
+Whenever possible, CockroachDB will [auto-retry a transaction internally](transactions.html#automatic-retries) without notifying the client. CockroachDB will only send a serialization error to the client when it cannot resolve the error automatically without client-side intervention.
 
-1. Design your schema and queries to reduce contention. For more information about how contention occurs and how to avoid it, see [Transaction Contention](performance-best-practices-overview.html#transaction-contention). In particular, if you are able to [send all of the statements in your transaction in a single batch](transactions.html#batched-statements), CockroachDB can usually automatically retry the entire transaction for you.
+The main reason why CockroachDB cannot auto-retry every serialization error without sending an error to the client is that the SQL language is "conversational" by design. The client can send arbitrary statements to the server during a transaction, receive some results, and then decide to issue other arbitrary statements inside the same transaction based on the server's response.
 
-1. Use historical reads with [`SELECT ... AS OF SYSTEM TIME`](as-of-system-time.html).
+Suppose that the client is [a Java application using JDBC](build-a-java-app-with-cockroachdb.html), or an analyst typing [`BEGIN`](begin-transaction.html) directly to [a SQL shell](cockroach-sql.html). In either case, the client is free to issue a `BEGIN`, wait an arbitrary amount of time, and issue additional statements. Meanwhile, other transactions are being processed by the system, potentially writing to the same data.
 
-1. In some cases, [`SELECT FOR UPDATE`](select-for-update.html) can also be used to reduce the occurrence of serialization errors.
+This "conversational" design means that there is no way for the server to always retry the arbitrary statements sent so far inside an open transaction. If there are different results for any given statement than there were at an earlier point in the currently open transaction's lifetime (likely due to the operations of other, concurrently-executing transactions), CockroachDB must defer to the client to decide how to handle that situation. This is why we recommend [keeping transactions as small as possible](performance-best-practices-overview.html#transaction-contention).
```

Comment:
Re-reading this section, does this seem out of place? It's an action but also links out to an overview doc. Maybe just remove it?

>This is why we recommend [keeping transactions as small as possible](https://deploy-preview-16280--cockroachdb-docs.netlify.app/docs/v23.1/performance-best-practices-overview#transaction-contention).

Reply:
This is under the `Overview` section: it is indeed intended as an overview. If you want this page 100% "do as we tell you" I can indeed remove the entire `## Overview` section. Isn't it helpful to be aware of serializable ordering always, auto-rety txns whenever possible, etc. I think there is some merit to providing this background before then proceeding to use these same concepts in actionable guidance throughout the rest of the page.

I've deleted just the "Suppose that a client is ... This is why we recommend keeping transactions as small as possible." section. What do you think? (I think it's still valuable to keep a brief overview of the concepts we use throughout the rest of the page).

Reply:
Oh yeah! Agree to keep the Overview section. I was just commenting on "This is why we recommend keeping transactions as small as possible." It felt random to have an action in this part of the doc.